vcs/util/tracer/tracer.go

```diff
@@ -0,0 +1,132 @@
+package tracer
+
+import (
+	"fmt"
+	"time"
+
+	"sourcegraph.com/sourcegraph/appdash"
+	"sourcegraph.com/sourcegraph/go-vcs/vcs"
+)
+
+func init() { appdash.RegisterEvent(GoVCS{}) }
+
+// GoVCS records a go-vcs method invocation.
+type GoVCS struct {
+	Name, Args string
+
+	StartTime time.Time
+	EndTime   time.Time
+}
+
+// Schema returns the constant "GoVCS".
+func (GoVCS) Schema() string { return "GoVCS" }
+
+func (e GoVCS) Start() time.Time { return e.StartTime }
+func (e GoVCS) End() time.Time   { return e.EndTime }
+
+// Wrap wraps the given VCS repository, returning a repository which emits
+// tracing events.
+func Wrap(r vcs.Repository, rec *appdash.Recorder) vcs.Repository {
+	// Wrap the repository.
+	t := repository{r: r, rec: rec}
+
+	// Also wrap optional interface.
+	final := vcs.Repository(t)
+	if b, ok := r.(vcs.Blamer); ok {
```

Comment:
It's possible I've misunderstood your comment, but I think a type switch cannot work here.

It is the case that `r` could implement all of these interface types (`Blamer`, `Differ`, `CrossRepoDiffer` and `FileLister`). A type switch would match only one of these, whereas the type assertions can match all of them in order.

The goal is to build and return a `vcs.Repository` which wraps all of these optional interfaces only if the input `vcs.Repository` also implements the interfaces.